// @progress/kendo-angular-scheduler/dist/es/localization/custom-messages.component.js
import * as tslib_1 from "tslib";
import { Component, forwardRef } from '@angular/core';
import { LocalizationService } from '@progress/kendo-angular-l10n';
import { Messages } from './messages';
/**
 * Custom component messages override default component messages.
 */
var SchedulerCustomMessagesComponent = /** @class */ (function (_super) {
    tslib_1.__extends(SchedulerCustomMessagesComponent, _super);
    function SchedulerCustomMessagesComponent(service) {
        var _this = _super.call(this) || this;
        _this.service = service;
        return _this;
    }
    Object.defineProperty(SchedulerCustomMessagesComponent.prototype, "override", {
        get: function () {
            return true;
        },
        enumerable: true,
        configurable: true
    });
    SchedulerCustomMessagesComponent.decorators = [
        { type: Component, args: [{
                    providers: [
                        {
                            provide: Messages,
                            useExisting: forwardRef(function () { return SchedulerCustomMessagesComponent; }) // tslint:disable-line:no-forward-ref
                        }
                    ],
                    selector: 'kendo-scheduler-messages',
                    template: ""
                },] },
    ];
    /** @nocollapse */
    SchedulerCustomMessagesComponent.ctorParameters = function () { return [
        { type: LocalizationService, },
    ]; };
    return SchedulerCustomMessagesComponent;
}(Messages));
export { SchedulerCustomMessagesComponent };
