// @progress/kendo-angular-editor/dist/es2015/dialogs/image-dialog.component.js
import { Component, Input } from '@angular/core';
import { FormControl, FormGroup, Validators } from '@angular/forms';
import { DialogRef, DialogContentBase } from '@progress/kendo-angular-dialog';
import { getNodeFromSelection } from '../prosemirror-utils';
import { safeString } from '../util';
/**
 * @hidden
 */
export class ImageDialogComponent extends DialogContentBase {
    constructor(dialog) {
        super(dialog);
        this.dialog = dialog;
        this.data = {
            src: '',
            alt: '',
            width: '',
            height: ''
        };
        this.src = new FormControl('', Validators.required);
        this.alt = new FormControl('');
        this.width = new FormControl('', Validators.min(1));
        this.height = new FormControl('', Validators.min(1));
        this.imageData = new FormGroup({
            src: this.src,
            alt: this.alt,
            width: this.width,
            height: this.height
        });
    }
    onCancelAction() {
        this.dialog.close();
    }
    onConfirmAction() {
        if (this.src.value) {
            this.editor.exec('insertImage', this.getData());
            this.dialog.close();
            this.editor.focus();
        }
    }
    setData(state) {
        const node = getNodeFromSelection(state);
        if (node) {
            this.src.patchValue(node.attrs.src);
            this.alt.patchValue(node.attrs.alt);
            this.width.patchValue(node.attrs.width);
            this.height.patchValue(node.attrs.height);
        }
    }
    getData() {
        return {
            src: this.src.value,
            alt: this.alt.value,
            width: this.normalizeDimension(this.width.value),
            height: this.normalizeDimension(this.height.value)
        };
    }
    normalizeDimension(value) {
        return Number.isNaN(parseInt(value, 10)) || parseInt(value, 10) <= 0 ? '' : safeString(parseInt(value, 10));
    }
}
ImageDialogComponent.decorators = [
    { type: Component, args: [{
                template: `
        <kendo-dialog-titlebar>
            Insert Image
        </kendo-dialog-titlebar>
        <div class="k-editor-dialog">
            <div class="k-editor-dialog k-popup-edit-form k-window-content k-content" >
                <div class="k-edit-form-container k-window-content">
                    <div class="k-edit-label">
                        <label for="k-editor-image-url">Web address</label>
                    </div>
                    <div class="k-edit-field">
                        <input [formControl]="src" type="text" class="k-input k-textbox">
                    </div>
                    <div class="k-edit-label">
                        <label for="k-editor-image-text">Alternate text</label>
                    </div>
                    <div class="k-edit-field">
                        <input [formControl]="alt" type="text" class="k-input k-textbox">
                    </div>
                    <div class="k-edit-label">
                        <label for="k-editor-image-width">Width (px)</label>
                    </div>
                    <div class="k-edit-field">
                        <input [formControl]="width" type="text" class="k-input k-textbox">
                    </div>
                    <div class="k-edit-label">
                        <label for="k-editor-image-height">Height (px)</label>
                    </div>
                    <div class="k-edit-field">
                        <input [formControl]="height" type="text" class="k-input k-textbox">
                    </div>
                </div>
            </div>
        </div>
        <kendo-dialog-actions>
            <button kendoButton (click)="onCancelAction()">Cancel</button>
            <button kendoButton [disabled]="imageData.invalid" (click)="onConfirmAction()" [primary]="true">Insert</button>
        </kendo-dialog-actions>
    `
            },] },
];
/** @nocollapse */
ImageDialogComponent.ctorParameters = () => [
    { type: DialogRef, },
];
ImageDialogComponent.propDecorators = {
    'editor': [{ type: Input },],
};
