// @progress/kendo-angular-editor/dist/es2015/dialogs/link-dialog.component.js
import { Component, ViewChild, Input } from '@angular/core';
import { DialogRef, DialogContentBase } from '@progress/kendo-angular-dialog';
import { schema } from '../config/schema';
import { getMark } from '../prosemirror-utils';
/**
 * @hidden
 */
export class LinkDialogComponent extends DialogContentBase {
    constructor(dialog) {
        super(dialog);
        this.dialog = dialog;
        this.data = { href: '', title: '', target: undefined };
    }
    onCancelAction() {
        this.dialog.close();
    }
    onConfirmAction() {
        this.editor.exec('createLink', this.getData());
        this.dialog.close();
        this.editor.focus();
    }
    setData(state) {
        const mark = getMark(state, schema.marks.link);
        if (mark) {
            this.data = {
                href: mark.attrs.href,
                title: mark.attrs.title,
                target: mark.attrs.target
            };
        }
    }
    getData() {
        return {
            href: this.href.nativeElement.value || undefined,
            title: this.title.nativeElement.value || undefined,
            target: this.target.nativeElement.checked ? '_blank' : undefined
        };
    }
}
LinkDialogComponent.decorators = [
    { type: Component, args: [{
                template: `
        <kendo-dialog-titlebar>
            Insert Link
        </kendo-dialog-titlebar>
        <div class="k-editor-dialog">
            <div class="k-editor-dialog k-popup-edit-form k-window-content k-content" >
                <div class="k-edit-form-container k-window-content">
                    <div class="k-edit-label">
                        <label for="k-editor-link-url">Web address</label>
                    </div>
                    <div class="k-edit-field">
                        <input #href [value]="data.href" type="text" class="k-input k-textbox">
                    </div>
                    <div class="k-edit-label k-editor-link-text-row">
                        <label for="k-editor-link-text">Title</label>
                    </div>
                    <div class="k-edit-field k-editor-link-text-row">
                        <input #title [value]="data.title" type="text" class="k-input k-textbox">
                    </div>
                    <div class="k-edit-field">
                        <input #target type="checkbox" class="k-checkbox">
                        <label for="k-editor-link-target" class="k-checkbox-label">Open link in new window</label>
                    </div>
                </div>
            </div>
        </div>
        <kendo-dialog-actions>
            <button kendoButton (click)="onCancelAction()">Cancel</button>
            <button kendoButton (click)="onConfirmAction()" [primary]="true">Insert</button>
        </kendo-dialog-actions>
    `
            },] },
];
/** @nocollapse */
LinkDialogComponent.ctorParameters = () => [
    { type: DialogRef, },
];
LinkDialogComponent.propDecorators = {
    'editor': [{ type: Input },],
    'href': [{ type: ViewChild, args: ['href',] },],
    'title': [{ type: ViewChild, args: ['title',] },],
    'target': [{ type: ViewChild, args: ['target',] },],
};
