// @progress/kendo-angular-editor/dist/es2015/dialogs/source-dialog.component.js
import { Component, ViewChild, Input } from '@angular/core';
// import { EditorService } from '../editor.service';
import { DialogRef, DialogContentBase } from '@progress/kendo-angular-dialog';
/**
 * @hidden
 */
export class SourceDialogComponent extends DialogContentBase {
    constructor(dialog) {
        super(dialog);
        this.dialog = dialog;
        this.data = '';
    }
    onCancelAction() {
        this.dialog.close();
    }
    onConfirmAction() {
        this.editor.exec('setHTML', this.getData());
        this.dialog.close();
        this.editor.focus();
    }
    getData() {
        return this.textarea.nativeElement.value;
    }
    setData() {
        this.data = this.indent(this.editor.getSource());
    }
    indent(content) {
        return content.replace(/<\/(p|li|ul|ol|h[1-6]|table|tr|td|th)>/ig, "</$1>\n")
            .replace(/<(ul|ol)([^>]*)><li/ig, "<$1$2>\n<li")
            .replace(/<br \/>/ig, "<br />\n")
            .replace(/\n$/, "");
    }
}
SourceDialogComponent.decorators = [
    { type: Component, args: [{
                template: `
        <kendo-dialog-titlebar>
            View Source
        </kendo-dialog-titlebar>
        <textarea [value]="data" #textarea class="k-editor-textarea k-input"></textarea>
        <kendo-dialog-actions>
            <button kendoButton (click)="onCancelAction()">Cancel</button>
            <button kendoButton (click)="onConfirmAction()" [primary]="true">Update</button>
        </kendo-dialog-actions>
    `,
                styles: [`
        >>> .k-editor-textarea {
            height: 100%;
        }
    `]
            },] },
];
/** @nocollapse */
SourceDialogComponent.ctorParameters = () => [
    { type: DialogRef, },
];
SourceDialogComponent.propDecorators = {
    'editor': [{ type: Input },],
    'textarea': [{ type: ViewChild, args: ['textarea',] },],
};
