// @progress/kendo-angular-scheduler/dist/es2015/views/day-time/day-time-view-item.component.js
import { Component, ElementRef, Renderer2, Input } from '@angular/core';
import { BaseViewItem } from '../view-items/base-view-item';
import { DayTimeSlotService } from '../day-time/day-time-slot.service';
import { IntlService } from '@progress/kendo-angular-intl';
import { LocalizationService } from '@progress/kendo-angular-l10n';
import { toLocalDate } from '@progress/kendo-date-math';
/**
 * @hidden
 */
export class DayTimeViewItemComponent extends BaseViewItem {
    constructor(intlService, slotService, localization, element, renderer) {
        super(slotService, localization, element, renderer);
        this.intlService = intlService;
    }
    get eventTitle() {
        return `(${this.eventTime}): ${this.item.event.title}`;
    }
    get eventTime() {
        return this.isAllDay ? this.intlService.toString(this.item.start.toLocalDate(), 't') :
            this.intlService.format('{0:t} - {1:t}', toLocalDate(this.item.startTime), toLocalDate(this.item.endTime));
    }
}
DayTimeViewItemComponent.decorators = [
    { type: Component, args: [{
                // tslint:disable-next-line:component-selector
                selector: '[dayTimeViewItem]',
                template: `
        <span class="k-event-actions">
            <span class="k-icon k-i-arrow-60-left" *ngIf="item.tail"></span>
            <span class="k-icon k-i-reload" *ngIf="isRecurrence"></span>
            <span class="k-icon k-i-non-recurrence" *ngIf="isRecurrenceException"></span>
        </span>
        <ng-container *ngIf="eventTemplate" [ngTemplateOutlet]="eventTemplate"
            [ngTemplateOutletContext]="{ $implicit: item.event, event: item.event, resources: resources }">
        </ng-container>
        <div *ngIf="!eventTemplate" [attr.title]="eventTitle">
            <div class="k-event-template k-event-time" *ngIf="!isAllDay">{{ eventTime }}</div>
            <div class="k-event-template">{{ item.event.title }}</div>
        </div>

        <span class="k-event-actions">
            <a href="#" *ngIf="removable" class="k-link k-event-delete" [attr.title]="deleteMessage" [attr.aria-label]="deleteMessage">
                <span class="k-icon k-i-close"></span>
            </a>
            <span class="k-icon k-i-arrow-60-right" *ngIf="item.head"></span>
        </span>

        <ng-container *ngIf="resizable && vertical">
            <span class="k-resize-handle k-resize-n"></span>
            <span class="k-resize-handle k-resize-s"></span>
        </ng-container>

        <ng-container *ngIf="resizable && !vertical">
            <span class="k-resize-handle k-resize-w"></span>
            <span class="k-resize-handle k-resize-e"></span>
        </ng-container>
    `
            },] },
];
/** @nocollapse */
DayTimeViewItemComponent.ctorParameters = () => [
    { type: IntlService, },
    { type: DayTimeSlotService, },
    { type: LocalizationService, },
    { type: ElementRef, },
    { type: Renderer2, },
];
DayTimeViewItemComponent.propDecorators = {
    'vertical': [{ type: Input },],
    'isAllDay': [{ type: Input },],
};
