// @progress/kendo-angular-editor/dist/es2015/tools/editor-button.directive.js
import { Directive, Input, Host } from '@angular/core';
import { ToolBarButtonComponent } from '@progress/kendo-angular-toolbar';
import { commandIcons } from '../config/command-icons';
import { EditorComponent } from '../editor.component';
/**
 * A directive which configures an existing `ToolBarButtonComponent` as an Editor tool
 * ([see example]({% slug toolbartools_editor %}#toc-built-in-tools)). According to the command that
 * is passed to it, `kendoEditorButton` predefines the `icon` and `click` event handlers of the button.
 * In addition, the directive updates the `selected` or `disabled` state of the button while the user is editing.
 *
 * @example
 * ```ts-no-run
 * <kendo-toolbar-button kendoEditorButton="bold"></kendo-toolbar-button>
 * ```
 *
 * The following example demonstrates how to change the default icon applied by the directive.
 *
 * @example
 * ```ts-no-run
 * <kendo-toolbar-button kendoEditorButton="bold" icon="blogger"></kendo-toolbar-button>
 * ```
 */
export class EditorButtonDirective {
    constructor(button, editor) {
        this.button = button;
        this.editor = editor;
    }
    ngOnInit() {
        if (!this.button.icon) {
            this.button.icon = commandIcons[this.command];
        }
        this.subs = this.button.click.subscribe(() => {
            this.editor.exec(this.command);
            this.editor.focus();
        });
        this.subs.add(this.editor.stateChange.subscribe((state) => {
            if (this.command === 'indent' || this.command === 'outdent') {
                this.button.disabled = !state[this.command];
            }
            else if (this.command === 'unlink') {
                this.button.disabled = !state.link;
            }
            else if (this.command === 'undo' || this.command === 'redo') {
                this.button.disabled = state[this.command];
            }
            else {
                this.button.selected = state[this.command];
            }
        }));
    }
    ngOnDestroy() {
        this.subs.unsubscribe();
    }
}
EditorButtonDirective.decorators = [
    { type: Directive, args: [{
                selector: 'kendo-toolbar-button[kendoEditorButton]'
            },] },
];
/** @nocollapse */
EditorButtonDirective.ctorParameters = () => [
    { type: ToolBarButtonComponent, },
    { type: EditorComponent, decorators: [{ type: Host },] },
];
EditorButtonDirective.propDecorators = {
    'command': [{ type: Input, args: ['kendoEditorButton',] },],
};
