// @progress/kendo-angular-editor/dist/es2015/tools/toolbar-dropdownlist.component.js
import { Component, Input, ViewChild, forwardRef, Output, EventEmitter } from '@angular/core';
import { ToolBarToolComponent } from '@progress/kendo-angular-toolbar';
import { outerWidth } from '../util';
/**
 * @hidden
 */
export class ToolBarDropDownListComponent extends ToolBarToolComponent {
    constructor() {
        super(...arguments);
        this.valueChange = new EventEmitter();
    }
    get outerWidth() {
        if (this.element) {
            return outerWidth(this.element.nativeElement);
        }
    }
}
ToolBarDropDownListComponent.decorators = [
    { type: Component, args: [{
                // tslint:disable-next-line:no-forward-ref
                providers: [{ provide: ToolBarToolComponent, useExisting: forwardRef(() => ToolBarDropDownListComponent) }],
                selector: 'kendo-toolbar-dropdownlist',
                template: `
        <ng-template #toolbarTemplate>
            <kendo-dropdownlist #element
                [defaultItem]="defaultItem"
                [textField]="textField"
                [valueField]="valueField"
                [data]="data"
                [value]="value"
                [valuePrimitive]="valuePrimitive"
                [itemDisabled]="itemDisabled"
                (valueChange)="valueChange.emit($event)"
            >
            </kendo-dropdownlist>
        </ng-template>
    `
            },] },
];
/** @nocollapse */
ToolBarDropDownListComponent.ctorParameters = () => [];
ToolBarDropDownListComponent.propDecorators = {
    'textField': [{ type: Input },],
    'valueField': [{ type: Input },],
    'valuePrimitive': [{ type: Input },],
    'value': [{ type: Input },],
    'data': [{ type: Input },],
    'defaultItem': [{ type: Input },],
    'itemDisabled': [{ type: Input },],
    'valueChange': [{ type: Output },],
    'toolbarTemplate': [{ type: ViewChild, args: ['toolbarTemplate',] },],
    'element': [{ type: ViewChild, args: ['element',] },],
};
