// @progress/kendo-angular-editor/dist/npm/tools/editor-button.directive.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var kendo_angular_toolbar_1 = require("@progress/kendo-angular-toolbar");
var command_icons_1 = require("../config/command-icons");
var editor_component_1 = require("../editor.component");
/**
 * A directive which configures an existing `ToolBarButtonComponent` as an Editor tool
 * ([see example]({% slug toolbartools_editor %}#toc-built-in-tools)). According to the command that
 * is passed to it, `kendoEditorButton` predefines the `icon` and `click` event handlers of the button.
 * In addition, the directive updates the `selected` or `disabled` state of the button while the user is editing.
 *
 * @example
 * ```ts-no-run
 * <kendo-toolbar-button kendoEditorButton="bold"></kendo-toolbar-button>
 * ```
 *
 * The following example demonstrates how to change the default icon applied by the directive.
 *
 * @example
 * ```ts-no-run
 * <kendo-toolbar-button kendoEditorButton="bold" icon="blogger"></kendo-toolbar-button>
 * ```
 */
var EditorButtonDirective = /** @class */ (function () {
    function EditorButtonDirective(button, editor) {
        this.button = button;
        this.editor = editor;
    }
    EditorButtonDirective.prototype.ngOnInit = function () {
        var _this = this;
        if (!this.button.icon) {
            this.button.icon = command_icons_1.commandIcons[this.command];
        }
        this.subs = this.button.click.subscribe(function () {
            _this.editor.exec(_this.command);
            _this.editor.focus();
        });
        this.subs.add(this.editor.stateChange.subscribe(function (state) {
            if (_this.command === 'indent' || _this.command === 'outdent') {
                _this.button.disabled = !state[_this.command];
            }
            else if (_this.command === 'unlink') {
                _this.button.disabled = !state.link;
            }
            else if (_this.command === 'undo' || _this.command === 'redo') {
                _this.button.disabled = state[_this.command];
            }
            else {
                _this.button.selected = state[_this.command];
            }
        }));
    };
    EditorButtonDirective.prototype.ngOnDestroy = function () {
        this.subs.unsubscribe();
    };
    EditorButtonDirective.decorators = [
        { type: core_1.Directive, args: [{
                    selector: 'kendo-toolbar-button[kendoEditorButton]'
                },] },
    ];
    /** @nocollapse */
    EditorButtonDirective.ctorParameters = function () { return [
        { type: kendo_angular_toolbar_1.ToolBarButtonComponent, },
        { type: editor_component_1.EditorComponent, decorators: [{ type: core_1.Host },] },
    ]; };
    EditorButtonDirective.propDecorators = {
        'command': [{ type: core_1.Input, args: ['kendoEditorButton',] },],
    };
    return EditorButtonDirective;
}());
exports.EditorButtonDirective = EditorButtonDirective;
