// @progress/kendo-angular-scheduler/dist/es/views/view-context.service.js
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs/BehaviorSubject';
import { Subject } from 'rxjs/Subject';
/**
 * A service which publishes information from the Scheduler to the views.
 * Views subscribe to changes in the context (selected date, event, and resource data) through this service.
 */
var ViewContextService = /** @class */ (function () {
    function ViewContextService() {
        this.actionSource = new Subject();
        this.itemsSource = new BehaviorSubject([]);
        this.selectedDateSource = new BehaviorSubject(null);
        this.resizeSource = new Subject();
        this.optionsChangeSource = new BehaviorSubject({});
        this.action = this.actionSource.asObservable();
        this.items = this.itemsSource.asObservable();
        this.selectedDate = this.selectedDateSource.asObservable();
        this.resize = this.resizeSource.asObservable();
        this.optionsChange = this.optionsChangeSource.asObservable();
    }
    /**
     * An internal method which is used by the Scheduler to publish unhandled navigation actions.
     *
     * @hidden
     */
    ViewContextService.prototype.notifyAction = function (action) {
        this.actionSource.next(action);
    };
    /**
     * An internal method which is used by the Scheduler to publish the current set of items.
     *
     * @hidden
     */
    ViewContextService.prototype.notifyItems = function (items) {
        this.itemsSource.next(items);
    };
    /**
     * An internal method which is used by the Scheduler to publish the currently selected date.
     *
     * @hidden
     */
    ViewContextService.prototype.notifySelectedDate = function (date) {
        this.selectedDateSource.next(date);
    };
    /**
     * An internal method which is used by the Scheduler to notify that the size changed.
     *
     * @hidden
     */
    ViewContextService.prototype.notifyResize = function () {
        this.resizeSource.next();
    };
    /**
     * An internal method which is used by the Scheduler to notify that the options changed.
     *
     * @hidden
     */
    ViewContextService.prototype.notifyOptionsChange = function (changes) {
        this.optionsChangeSource.next(changes);
    };
    ViewContextService.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    ViewContextService.ctorParameters = function () { return []; };
    return ViewContextService;
}());
export { ViewContextService };
