// @progress/kendo-angular-editor/dist/npm/dialogs/image-dialog.component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
var forms_1 = require("@angular/forms");
var kendo_angular_dialog_1 = require("@progress/kendo-angular-dialog");
var prosemirror_utils_1 = require("../prosemirror-utils");
var util_1 = require("../util");
/**
 * @hidden
 */
var ImageDialogComponent = /** @class */ (function (_super) {
    tslib_1.__extends(ImageDialogComponent, _super);
    function ImageDialogComponent(dialog) {
        var _this = _super.call(this, dialog) || this;
        _this.dialog = dialog;
        _this.data = {
            src: '',
            alt: '',
            width: '',
            height: ''
        };
        _this.src = new forms_1.FormControl('', forms_1.Validators.required);
        _this.alt = new forms_1.FormControl('');
        _this.width = new forms_1.FormControl('', forms_1.Validators.min(1));
        _this.height = new forms_1.FormControl('', forms_1.Validators.min(1));
        _this.imageData = new forms_1.FormGroup({
            src: _this.src,
            alt: _this.alt,
            width: _this.width,
            height: _this.height
        });
        return _this;
    }
    ImageDialogComponent.prototype.onCancelAction = function () {
        this.dialog.close();
    };
    ImageDialogComponent.prototype.onConfirmAction = function () {
        if (this.src.value) {
            this.editor.exec('insertImage', this.getData());
            this.dialog.close();
            this.editor.focus();
        }
    };
    ImageDialogComponent.prototype.setData = function (state) {
        var node = prosemirror_utils_1.getNodeFromSelection(state);
        if (node) {
            this.src.patchValue(node.attrs.src);
            this.alt.patchValue(node.attrs.alt);
            this.width.patchValue(node.attrs.width);
            this.height.patchValue(node.attrs.height);
        }
    };
    ImageDialogComponent.prototype.getData = function () {
        return {
            src: this.src.value,
            alt: this.alt.value,
            width: this.normalizeDimension(this.width.value),
            height: this.normalizeDimension(this.height.value)
        };
    };
    ImageDialogComponent.prototype.normalizeDimension = function (value) {
        return Number.isNaN(parseInt(value, 10)) || parseInt(value, 10) <= 0 ? '' : util_1.safeString(parseInt(value, 10));
    };
    ImageDialogComponent.decorators = [
        { type: core_1.Component, args: [{
                    template: "\n        <kendo-dialog-titlebar>\n            Insert Image\n        </kendo-dialog-titlebar>\n        <div class=\"k-editor-dialog\">\n            <div class=\"k-editor-dialog k-popup-edit-form k-window-content k-content\" >\n                <div class=\"k-edit-form-container k-window-content\">\n                    <div class=\"k-edit-label\">\n                        <label for=\"k-editor-image-url\">Web address</label>\n                    </div>\n                    <div class=\"k-edit-field\">\n                        <input [formControl]=\"src\" type=\"text\" class=\"k-input k-textbox\">\n                    </div>\n                    <div class=\"k-edit-label\">\n                        <label for=\"k-editor-image-text\">Alternate text</label>\n                    </div>\n                    <div class=\"k-edit-field\">\n                        <input [formControl]=\"alt\" type=\"text\" class=\"k-input k-textbox\">\n                    </div>\n                    <div class=\"k-edit-label\">\n                        <label for=\"k-editor-image-width\">Width (px)</label>\n                    </div>\n                    <div class=\"k-edit-field\">\n                        <input [formControl]=\"width\" type=\"text\" class=\"k-input k-textbox\">\n                    </div>\n                    <div class=\"k-edit-label\">\n                        <label for=\"k-editor-image-height\">Height (px)</label>\n                    </div>\n                    <div class=\"k-edit-field\">\n                        <input [formControl]=\"height\" type=\"text\" class=\"k-input k-textbox\">\n                    </div>\n                </div>\n            </div>\n        </div>\n        <kendo-dialog-actions>\n            <button kendoButton (click)=\"onCancelAction()\">Cancel</button>\n            <button kendoButton [disabled]=\"imageData.invalid\" (click)=\"onConfirmAction()\" [primary]=\"true\">Insert</button>\n        </kendo-dialog-actions>\n    "
                },] },
    ];
    /** @nocollapse */
    ImageDialogComponent.ctorParameters = function () { return [
        { type: kendo_angular_dialog_1.DialogRef, },
    ]; };
    ImageDialogComponent.propDecorators = {
        'editor': [{ type: core_1.Input },],
    };
    return ImageDialogComponent;
}(kendo_angular_dialog_1.DialogContentBase));
exports.ImageDialogComponent = ImageDialogComponent;
