// @progress/kendo-angular-scheduler/dist/npm/views/templates.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var agenda_date_template_directive_1 = require("./templates/agenda-date-template.directive");
exports.AgendaDateTemplateDirective = agenda_date_template_directive_1.AgendaDateTemplateDirective;
var agenda_time_template_directive_1 = require("./templates/agenda-time-template.directive");
exports.AgendaTimeTemplateDirective = agenda_time_template_directive_1.AgendaTimeTemplateDirective;
var all_day_event_template_directive_1 = require("./templates/all-day-event-template.directive");
exports.AllDayEventTemplateDirective = all_day_event_template_directive_1.AllDayEventTemplateDirective;
var all_day_slot_template_directive_1 = require("./templates/all-day-slot-template.directive");
exports.AllDaySlotTemplateDirective = all_day_slot_template_directive_1.AllDaySlotTemplateDirective;
var date_header_template_directive_1 = require("./templates/date-header-template.directive");
exports.DateHeaderTemplateDirective = date_header_template_directive_1.DateHeaderTemplateDirective;
var event_template_directive_1 = require("./templates/event-template.directive");
exports.EventTemplateDirective = event_template_directive_1.EventTemplateDirective;
var group_header_template_directive_1 = require("./templates/group-header-template.directive");
exports.GroupHeaderTemplateDirective = group_header_template_directive_1.GroupHeaderTemplateDirective;
var major_time_header_template_directive_1 = require("./templates/major-time-header-template.directive");
exports.MajorTimeHeaderTemplateDirective = major_time_header_template_directive_1.MajorTimeHeaderTemplateDirective;
var minor_time_header_template_directive_1 = require("./templates/minor-time-header-template.directive");
exports.MinorTimeHeaderTemplateDirective = minor_time_header_template_directive_1.MinorTimeHeaderTemplateDirective;
var month_day_slot_template_directive_1 = require("./templates/month-day-slot-template.directive");
exports.MonthDaySlotTemplateDirective = month_day_slot_template_directive_1.MonthDaySlotTemplateDirective;
var time_slot_template_directive_1 = require("./templates/time-slot-template.directive");
exports.TimeSlotTemplateDirective = time_slot_template_directive_1.TimeSlotTemplateDirective;
