// @progress/kendo-angular-scheduler/dist/npm/views/view-context.service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var BehaviorSubject_1 = require("rxjs/BehaviorSubject");
var Subject_1 = require("rxjs/Subject");
/**
 * A service which publishes information from the Scheduler to the views.
 * Views subscribe to changes in the context (selected date, event, and resource data) through this service.
 */
var ViewContextService = /** @class */ (function () {
    function ViewContextService() {
        this.actionSource = new Subject_1.Subject();
        this.itemsSource = new BehaviorSubject_1.BehaviorSubject([]);
        this.selectedDateSource = new BehaviorSubject_1.BehaviorSubject(null);
        this.resizeSource = new Subject_1.Subject();
        this.optionsChangeSource = new BehaviorSubject_1.BehaviorSubject({});
        this.action = this.actionSource.asObservable();
        this.items = this.itemsSource.asObservable();
        this.selectedDate = this.selectedDateSource.asObservable();
        this.resize = this.resizeSource.asObservable();
        this.optionsChange = this.optionsChangeSource.asObservable();
    }
    /**
     * An internal method which is used by the Scheduler to publish unhandled navigation actions.
     *
     * @hidden
     */
    ViewContextService.prototype.notifyAction = function (action) {
        this.actionSource.next(action);
    };
    /**
     * An internal method which is used by the Scheduler to publish the current set of items.
     *
     * @hidden
     */
    ViewContextService.prototype.notifyItems = function (items) {
        this.itemsSource.next(items);
    };
    /**
     * An internal method which is used by the Scheduler to publish the currently selected date.
     *
     * @hidden
     */
    ViewContextService.prototype.notifySelectedDate = function (date) {
        this.selectedDateSource.next(date);
    };
    /**
     * An internal method which is used by the Scheduler to notify that the size changed.
     *
     * @hidden
     */
    ViewContextService.prototype.notifyResize = function () {
        this.resizeSource.next();
    };
    /**
     * An internal method which is used by the Scheduler to notify that the options changed.
     *
     * @hidden
     */
    ViewContextService.prototype.notifyOptionsChange = function (changes) {
        this.optionsChangeSource.next(changes);
    };
    ViewContextService.decorators = [
        { type: core_1.Injectable },
    ];
    /** @nocollapse */
    ViewContextService.ctorParameters = function () { return []; };
    return ViewContextService;
}());
exports.ViewContextService = ViewContextService;
