// @progress/kendo-angular-scheduler/dist/es/views/multi-day/utils.js
import { intersects, toUTCDate, roundAllDayEnd } from '../utils';
/** @hidden */
export var isMultiDay = function (_a) {
    var start = _a.start, end = _a.end;
    var startDate = start.stripTime();
    var endDate = end.stripTime();
    return startDate.getTime() !== endDate.getTime() &&
        (endDate.getTime() !== end.getTime() || startDate.addDays(1).getTime() !== endDate.getTime());
};
/** @hidden */
export var createTasks = function (periodStart, periodEnd, items) {
    var tasks = [];
    var utcStart = toUTCDate(periodStart);
    var utcEnd = toUTCDate(periodEnd);
    for (var index = 0; index < items.length; index++) {
        var item = items[index];
        var task = {
            index: index,
            start: item.start,
            end: item.end,
            event: item.event,
            isAllDay: item.event.isAllDay || isMultiDay(item)
        };
        var endTime = task.isAllDay ? roundAllDayEnd(task) : task.end;
        var startTime = task.isAllDay ? task.start.stripTime() : task.start;
        task.startTime = startTime.toUTCDate();
        task.endTime = endTime.toUTCDate();
        if (intersects(task.startTime, task.endTime, utcStart, utcEnd)) {
            tasks.push(task);
            task.head = task.endTime > utcEnd;
            task.tail = task.startTime < utcStart;
        }
    }
    return tasks;
};
