// @progress/kendo-angular-scheduler/dist/es/editing/dialogs.service.js
import { Injectable, ChangeDetectorRef } from '@angular/core';
import { DialogService, DialogCloseResult } from '@progress/kendo-angular-dialog';
import { LocalizationService } from '@progress/kendo-angular-l10n';
import { map } from 'rxjs/operators/map';
/**
 * @hidden
 */
var DialogsService = /** @class */ (function () {
    function DialogsService(dialogService, localization, changeDetector) {
        this.dialogService = dialogService;
        this.localization = localization;
        this.changeDetector = changeDetector;
    }
    DialogsService.prototype.openRemoveConfirmationDialog = function () {
        var dialog = this.dialogService.open({
            title: this.textFor('deleteDialogTitle'),
            content: this.textFor('deleteConfirmation'),
            actions: [
                { text: this.textFor('cancel'), value: false },
                { text: this.textFor('destroy'), value: true }
            ],
            appendTo: this.container
        });
        this.changeDetector.markForCheck();
        return dialog.result.pipe(map(function (result) {
            if (result instanceof DialogCloseResult) {
                return false;
            }
            var res = result;
            return res.value;
        }));
    };
    DialogsService.prototype.openRecurringConfirmationDialog = function (operation) {
        var dialog = this.dialogService.open({
            actions: [
                {
                    text: operation === 0 /* Edit */ ? this.textFor('editOccurrence') : this.textFor('deleteOccurrence'),
                    value: 1 /* Occurrence */
                },
                {
                    text: operation === 0 /* Edit */ ? this.textFor('editSeries') : this.textFor('deleteSeries'),
                    value: 2 /* Series */
                }
            ],
            appendTo: this.container,
            content: operation === 0 /* Edit */ ? this.textFor('editRecurringConfirmation') : this.textFor('deleteRecurringConfirmation'),
            title: operation === 0 /* Edit */ ? this.textFor('editRecurringDialogTitle') : this.textFor('deleteRecurringDialogTitle')
        });
        this.changeDetector.markForCheck();
        return dialog.result.pipe(map(function (result) {
            if (result instanceof DialogCloseResult) {
                return undefined;
            }
            var res = result;
            return res.value;
        }));
    };
    DialogsService.prototype.textFor = function (key) {
        return this.localization.get(key);
    };
    DialogsService.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    DialogsService.ctorParameters = function () { return [
        { type: DialogService, },
        { type: LocalizationService, },
        { type: ChangeDetectorRef, },
    ]; };
    return DialogsService;
}());
export { DialogsService };
