// @progress/kendo-angular-editor/dist/npm/editor.module.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var common_1 = require("@angular/common");
var forms_1 = require("@angular/forms");
var kendo_angular_toolbar_1 = require("@progress/kendo-angular-toolbar");
var kendo_angular_dropdowns_1 = require("@progress/kendo-angular-dropdowns");
var kendo_angular_dialog_1 = require("@progress/kendo-angular-dialog");
var kendo_angular_buttons_1 = require("@progress/kendo-angular-buttons");
var editor_component_1 = require("./editor.component");
var toolbar_dropdownlist_component_1 = require("./tools/toolbar-dropdownlist.component");
var editor_button_directive_1 = require("./tools/editor-button.directive");
var editor_dropdown_directive_1 = require("./tools/editor-dropdown.directive");
var toolbar_combobox_component_1 = require("./tools/toolbar-combobox.component");
var editor_dialog_directive_1 = require("./tools/editor-dialog.directive");
var link_dialog_component_1 = require("./dialogs/link-dialog.component");
var source_dialog_component_1 = require("./dialogs/source-dialog.component");
var image_dialog_component_1 = require("./dialogs/image-dialog.component");
var COMPONENT_DIRECTIVES = [
    editor_button_directive_1.EditorButtonDirective,
    editor_dropdown_directive_1.EditorDropDownDirective,
    editor_dialog_directive_1.EditorDialogDirective
];
var TOOLBAR_TOOLS = [
    toolbar_dropdownlist_component_1.ToolBarDropDownListComponent,
    toolbar_combobox_component_1.ToolBarComboBoxComponent
];
var TOOLBAR_DIALOGS = [
    link_dialog_component_1.LinkDialogComponent,
    source_dialog_component_1.SourceDialogComponent,
    image_dialog_component_1.ImageDialogComponent
];
/**
 * Represents the [NgModule]({{ site.data.urls.angular['ngmoduleapi'] }}) definition for the Editor component.
 *
 * The package exports:
 * - `EditorComponent`&mdash;The `EditorComponent` class.
 * - `EditorButtonDirective`&mdash;The `EditorButton` directive class.
 * - `EditorDropDownDirective`&mdash;The `EditorDropDown` directive class.
 * - `EditorDialogDirective`&mdash;The `EditorDialog` directive class.
 * - `ToolBarDropDownListComponent`&mdash;The `ToolBarDropDownListComponent` directive class.
 * - `ButtonModule`&mdash;The `KendoButton` module.
 * - `ToolBarModule`&mdash;The `KendoToolBar` module.
 *
 *  * @example
 *
 * ```ts-no-run
 * // Import the Editor module
 * import { EditorModule } from '@progress/kendo-angular-editor';
 *
 * // The browser platform with a compiler
 * import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
 *
 * import { NgModule } from '@angular/core';
 *
 * // Import the app component
 * import { AppComponent } from './app.component';
 *
 * // Define the app module
 * _@NgModule({
 *     declarations: [AppComponent], // declare the app component
 *     imports:      [BrowserModule, EditorModule], // import the Editor module
 *     bootstrap:    [AppComponent]
 * })
 * export class AppModule {}
 *
 * // Compile and launch the module
 * platformBrowserDynamic().bootstrapModule(AppModule);
 *
 * ```
 */
var EditorModule = /** @class */ (function () {
    function EditorModule() {
    }
    EditorModule.decorators = [
        { type: core_1.NgModule, args: [{
                    declarations: [
                        editor_component_1.EditorComponent,
                        COMPONENT_DIRECTIVES,
                        TOOLBAR_TOOLS,
                        TOOLBAR_DIALOGS
                    ],
                    exports: [
                        editor_component_1.EditorComponent,
                        COMPONENT_DIRECTIVES,
                        TOOLBAR_TOOLS,
                        kendo_angular_toolbar_1.ToolBarModule,
                        kendo_angular_buttons_1.ButtonModule,
                        forms_1.FormsModule,
                        forms_1.ReactiveFormsModule
                    ],
                    imports: [
                        common_1.CommonModule,
                        kendo_angular_toolbar_1.ToolBarModule,
                        kendo_angular_dropdowns_1.DropDownsModule,
                        kendo_angular_dialog_1.DialogModule,
                        kendo_angular_buttons_1.ButtonModule,
                        forms_1.FormsModule,
                        forms_1.ReactiveFormsModule
                    ],
                    entryComponents: [
                        TOOLBAR_DIALOGS
                    ]
                },] },
    ];
    /** @nocollapse */
    EditorModule.ctorParameters = function () { return []; };
    return EditorModule;
}());
exports.EditorModule = EditorModule;
