// @progress/kendo-angular-editor/dist/npm/util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * @hidden
 */
function outerWidth(element) {
    var width = element.offsetWidth;
    var style = getComputedStyle(element);
    width += (parseFloat(style.marginLeft) || 0 + parseFloat(style.marginRight) || 0);
    return width;
}
exports.outerWidth = outerWidth;
/**
 * @hidden
 */
exports.serializeDOMAttrs = function (el) {
    return Array.from(el.attributes)
        .reduce(function (acc, curr) {
        return Object.assign({}, acc, (_a = {}, _a[curr.name] = curr.value, _a));
        var _a;
    }, {});
};
/**
 * @hidden
 */
exports.removeEntries = function (obj, predicate) {
    return Object.keys(obj)
        .filter(function (key) { return predicate(key); })
        .reduce(function (acc, curr) {
        return Object.assign(acc, (_a = {}, _a[curr] = obj[curr], _a));
        var _a;
    }, {});
};
/**
 * @hidden
 */
exports.removeEmptyEntries = function (obj) {
    var predicate = function (key) { return obj[key] !== null && obj[key] !== undefined && obj[key] !== ''; };
    return exports.removeEntries(obj, predicate);
};
/**
 * @hidden
 */
exports.isEmpty = function (obj) { return Object.keys(obj).length === 0; };
/**
 * @hidden
 */
exports.isNullOrUndefined = function (value) { return value === undefined || value === null; };
/**
 * @hidden
 */
exports.safeString = function (value) { return exports.isNullOrUndefined(value) ? '' : value.toString(); };
/**
 * @hidden
 */
exports.first = function (arr) { return arr[0]; };
/**
 * @hidden
 */
exports.last = function (arr) { return arr[arr.length - 1]; };
/**
 * @hidden
 */
exports.unique = function (arr) { return Array.from(new Set(arr)); };
