// @progress/kendo-angular-scheduler/dist/es2015/views/view-items/base-view-item.js
import { Input, HostBinding } from '@angular/core';
import { anyChanged } from '../../common/changes';
import { isRecurrence, isRecurrenceException } from '../utils';
/**
 * @hidden
 */
export class BaseViewItem {
    constructor(slotService, localization, element, renderer) {
        this.slotService = slotService;
        this.localization = localization;
        this.element = element;
        this.renderer = renderer;
        this.className = true;
    }
    get taskIndex() {
        return this.item.index;
    }
    get deleteMessage() {
        return this.localization.get('deleteTitle');
    }
    get resizable() {
        return this.editable && this.editable.resize !== false;
    }
    get removable() {
        return this.editable && this.editable.remove !== false;
    }
    get resources() {
        return this.item.resources[0].resources;
    }
    get isRecurrence() {
        return isRecurrence(this.item);
    }
    get isRecurrenceException() {
        return isRecurrenceException(this.item);
    }
    get nativeElement() {
        if (this.element) {
            return this.element.nativeElement;
        }
    }
    setStyles(styles) {
        const element = this.nativeElement;
        if (element) {
            for (let name in styles) {
                if (styles.hasOwnProperty(name)) {
                    this.renderer.setStyle(element, name, styles[name]);
                }
            }
        }
    }
    toggle(visible) {
        this.setStyles({ display: visible ? 'block' : 'none' });
    }
    reflow() {
        const rect = this.rect;
        if (rect) {
            this.setStyles({
                left: !this.localization.rtl ? `${rect.left}px` : '',
                right: this.localization.rtl ? `${rect.left}px` : '',
                top: `${rect.top}px`,
                width: `${rect.width}px`,
                height: `${rect.height}px`,
                display: 'block'
            });
        }
    }
    ngOnInit() {
        if (this.dragHint) {
            return;
        }
        this.slotsChangeSubscription = this.slotService.slotsChange.subscribe(() => {
            this.rect = null;
            this.setStyles({
                display: 'none',
                width: 0,
                left: 0
            });
            this.slotService.unregisterItem(this, this.resourceIndex, this.index);
            this.slotService.registerItem(this);
        });
    }
    ngOnChanges(changes) {
        if (this.dragHint) {
            return;
        }
        if (anyChanged(['resourceIndex', 'index'], changes)) {
            const { resourceIndex, index } = changes;
            const previousResourceIndex = resourceIndex ? resourceIndex.previousValue : this.resourceIndex;
            const previousIndex = index ? index.previousValue : this.index;
            this.slotService.unregisterItem(this, previousResourceIndex, previousIndex);
            this.slotService.registerItem(this);
        }
    }
    ngOnDestroy() {
        if (this.dragHint) {
            return;
        }
        this.slotService.unregisterItem(this);
        if (this.slotsChangeSubscription) {
            this.slotsChangeSubscription.unsubscribe();
        }
    }
}
BaseViewItem.propDecorators = {
    'item': [{ type: Input },],
    'resourceIndex': [{ type: Input },],
    'index': [{ type: Input },],
    'eventTemplate': [{ type: Input },],
    'editable': [{ type: Input },],
    'dragHint': [{ type: Input },],
    'className': [{ type: HostBinding, args: ['class.k-event',] },],
    'taskIndex': [{ type: HostBinding, args: ['attr.data-task-index',] },],
};
