// @progress/kendo-angular-editor/dist/es/editor.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { ToolBarModule } from '@progress/kendo-angular-toolbar';
import { DropDownsModule } from '@progress/kendo-angular-dropdowns';
import { DialogModule } from '@progress/kendo-angular-dialog';
import { ButtonModule } from '@progress/kendo-angular-buttons';
import { EditorComponent } from './editor.component';
import { ToolBarDropDownListComponent } from './tools/toolbar-dropdownlist.component';
import { EditorButtonDirective } from './tools/editor-button.directive';
import { EditorDropDownDirective } from './tools/editor-dropdown.directive';
import { ToolBarComboBoxComponent } from './tools/toolbar-combobox.component';
import { EditorDialogDirective } from './tools/editor-dialog.directive';
import { LinkDialogComponent } from './dialogs/link-dialog.component';
import { SourceDialogComponent } from './dialogs/source-dialog.component';
import { ImageDialogComponent } from './dialogs/image-dialog.component';
var COMPONENT_DIRECTIVES = [
    EditorButtonDirective,
    EditorDropDownDirective,
    EditorDialogDirective
];
var TOOLBAR_TOOLS = [
    ToolBarDropDownListComponent,
    ToolBarComboBoxComponent
];
var TOOLBAR_DIALOGS = [
    LinkDialogComponent,
    SourceDialogComponent,
    ImageDialogComponent
];
/**
 * Represents the [NgModule]({{ site.data.urls.angular['ngmoduleapi'] }}) definition for the Editor component.
 *
 * The package exports:
 * - `EditorComponent`&mdash;The `EditorComponent` class.
 * - `EditorButtonDirective`&mdash;The `EditorButton` directive class.
 * - `EditorDropDownDirective`&mdash;The `EditorDropDown` directive class.
 * - `EditorDialogDirective`&mdash;The `EditorDialog` directive class.
 * - `ToolBarDropDownListComponent`&mdash;The `ToolBarDropDownListComponent` directive class.
 * - `ButtonModule`&mdash;The `KendoButton` module.
 * - `ToolBarModule`&mdash;The `KendoToolBar` module.
 *
 *  * @example
 *
 * ```ts-no-run
 * // Import the Editor module
 * import { EditorModule } from '@progress/kendo-angular-editor';
 *
 * // The browser platform with a compiler
 * import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
 *
 * import { NgModule } from '@angular/core';
 *
 * // Import the app component
 * import { AppComponent } from './app.component';
 *
 * // Define the app module
 * _@NgModule({
 *     declarations: [AppComponent], // declare the app component
 *     imports:      [BrowserModule, EditorModule], // import the Editor module
 *     bootstrap:    [AppComponent]
 * })
 * export class AppModule {}
 *
 * // Compile and launch the module
 * platformBrowserDynamic().bootstrapModule(AppModule);
 *
 * ```
 */
var EditorModule = /** @class */ (function () {
    function EditorModule() {
    }
    EditorModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        EditorComponent,
                        COMPONENT_DIRECTIVES,
                        TOOLBAR_TOOLS,
                        TOOLBAR_DIALOGS
                    ],
                    exports: [
                        EditorComponent,
                        COMPONENT_DIRECTIVES,
                        TOOLBAR_TOOLS,
                        ToolBarModule,
                        ButtonModule,
                        FormsModule,
                        ReactiveFormsModule
                    ],
                    imports: [
                        CommonModule,
                        ToolBarModule,
                        DropDownsModule,
                        DialogModule,
                        ButtonModule,
                        FormsModule,
                        ReactiveFormsModule
                    ],
                    entryComponents: [
                        TOOLBAR_DIALOGS
                    ]
                },] },
    ];
    /** @nocollapse */
    EditorModule.ctorParameters = function () { return []; };
    return EditorModule;
}());
export { EditorModule };
