// @progress/kendo-angular-scheduler/dist/es2015/editing/recurrence/recurrence-end-rule-editor.component.js
import { Component, ViewChildren } from '@angular/core';
import { RecurrenceService } from './recurrence.service';
import { isPresent } from '../../common/util';
import { LocalizationService } from '@progress/kendo-angular-l10n';
import { EndRuleRadioButtonDirective } from './end-rule-radio-button.directive';
/**
 * @hidden
 */
export class RecurrenceEndRuleEditorComponent {
    constructor(recurrence, localization) {
        this.recurrence = recurrence;
        this.localization = localization;
        this.tempCount = 1;
        this.setInitialValues();
        this.subscribeChanges();
    }
    ngOnDestroy() {
        if (this.subscriptions) {
            this.subscriptions.unsubscribe();
        }
    }
    setEndRule(endRule) {
        if (endRule === 'count') {
            this.recurrence.rrule.count = this.tempCount;
        }
        else if (endRule === 'until') {
            this.recurrence.until = this.tempUntil;
        }
    }
    get rrule() {
        return this.recurrence.rrule;
    }
    get countValue() {
        return this.rrule.count || this.tempCount;
    }
    get untilValue() {
        const currentStart = this.recurrence.start;
        const currentUntil = this.recurrence.until;
        if (isPresent(currentUntil)) {
            return currentUntil;
        }
        else if (isPresent(this.tempUntil)) {
            return this.tempUntil;
        }
        else {
            return new Date(currentStart.getFullYear(), currentStart.getMonth(), currentStart.getDate(), 23, 59, 59);
        }
    }
    get untilMin() {
        const currentStart = this.recurrence.start;
        const currentUntil = this.recurrence.until;
        if (isPresent(currentUntil) && currentUntil < currentStart) {
            return currentUntil;
        }
        else {
            return currentStart;
        }
    }
    get isCountDisabled() {
        return this.recurrence.endRule !== 'count';
    }
    get isUntilDisabled() {
        return this.recurrence.endRule !== 'until';
    }
    onCountChange(count) {
        this.recurrence.count = this.tempCount = count;
    }
    onUntilChange(until) {
        const newUntil = new Date(until.getFullYear(), until.getMonth(), until.getDate(), 23, 59, 59);
        this.recurrence.until = this.tempUntil = newUntil;
    }
    textFor(key) {
        return this.localization.get(key);
    }
    onEndLabelClick() {
        const selected = this.endRuleRadioButtons.toArray().find(r => r.elem.checked);
        if (selected) {
            selected.elem.focus();
        }
    }
    setInitialValues() {
        this.tempCount = this.rrule.count || 1;
        this.tempUntil = this.untilValue;
    }
    subscribeChanges() {
        this.subscriptions = this.recurrence.endRuleChange.subscribe((endRule) => {
            this.setEndRule(endRule);
        });
        this.subscriptions.add(this.recurrence.frequencyChange.subscribe(() => {
            this.setInitialValues();
        }));
    }
}
RecurrenceEndRuleEditorComponent.decorators = [
    { type: Component, args: [{
                selector: 'kendo-recurrence-end-rule-editor',
                template: `
        <div class='k-edit-label'>
            <label (click)="onEndLabelClick()">{{ textFor('endLabel') }}</label>
        </div>
        <div class='k-edit-field'>
            <ul class='k-reset'>
                <li>
                    <input [kendoRecurrenceEndRuleRadioButton]="'k-endrule-never'" />
                    <label class='k-radio-label' for='k-endrule-never'>{{ textFor('endNever') }}</label>
                </li>
                <li>
                    <input [kendoRecurrenceEndRuleRadioButton]="'k-endrule-count'" />
                    <label class='k-radio-label' for='k-endrule-count'>{{ textFor('endAfter') }}</label>

                    <kendo-numerictextbox
                        [style.width.px]='70'
                        [autoCorrect]='true'
                        [decimals]='0'
                        [disabled]='isCountDisabled'
                        [format]="'#'"
                        [min]='1'
                        [value]='countValue'
                        (valueChange)='onCountChange($event)'>
                    </kendo-numerictextbox>
                    <span>{{ textFor('endOccurrence') }}</span>
                </li>
                <li>
                    <input [kendoRecurrenceEndRuleRadioButton]="'k-endrule-until'" />
                    <label class='k-radio-label' for='k-endrule-until'>{{ textFor('endOn') }}</label>

                    <kendo-datepicker
                        [style.width.px]='150'
                        [disabled]='isUntilDisabled'
                        [min]='untilMin'
                        [value]='untilValue'
                        (valueChange)='onUntilChange($event)'>
                    </kendo-datepicker>
                </li>
            </ul>
        </div>
    `
            },] },
];
/** @nocollapse */
RecurrenceEndRuleEditorComponent.ctorParameters = () => [
    { type: RecurrenceService, },
    { type: LocalizationService, },
];
RecurrenceEndRuleEditorComponent.propDecorators = {
    'endRuleRadioButtons': [{ type: ViewChildren, args: [EndRuleRadioButtonDirective,] },],
};
