// @progress/kendo-angular-scheduler/dist/npm/views/multi-day/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var utils_1 = require("../utils");
/** @hidden */
exports.isMultiDay = function (_a) {
    var start = _a.start, end = _a.end;
    var startDate = start.stripTime();
    var endDate = end.stripTime();
    return startDate.getTime() !== endDate.getTime() &&
        (endDate.getTime() !== end.getTime() || startDate.addDays(1).getTime() !== endDate.getTime());
};
/** @hidden */
exports.createTasks = function (periodStart, periodEnd, items) {
    var tasks = [];
    var utcStart = utils_1.toUTCDate(periodStart);
    var utcEnd = utils_1.toUTCDate(periodEnd);
    for (var index = 0; index < items.length; index++) {
        var item = items[index];
        var task = {
            index: index,
            start: item.start,
            end: item.end,
            event: item.event,
            isAllDay: item.event.isAllDay || exports.isMultiDay(item)
        };
        var endTime = task.isAllDay ? utils_1.roundAllDayEnd(task) : task.end;
        var startTime = task.isAllDay ? task.start.stripTime() : task.start;
        task.startTime = startTime.toUTCDate();
        task.endTime = endTime.toUTCDate();
        if (utils_1.intersects(task.startTime, task.endTime, utcStart, utcEnd)) {
            tasks.push(task);
            task.head = task.endTime > utcEnd;
            task.tail = task.startTime < utcStart;
        }
    }
    return tasks;
};
