// @progress/kendo-angular-scheduler/dist/npm/views/agenda/agenda-view-list.component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var kendo_angular_intl_1 = require("@progress/kendo-angular-intl");
var kendo_angular_l10n_1 = require("@progress/kendo-angular-l10n");
var kendo_date_math_1 = require("@progress/kendo-date-math");
/**
 * @hidden
 */
var AgendaListComponent = /** @class */ (function () {
    function AgendaListComponent(intlService, localization) {
        this.intlService = intlService;
        this.localization = localization;
        this.classes = true;
    }
    AgendaListComponent.prototype.extractDataItem = function (item) {
        return item.type === "group" ? item.dataItem.items[0] : item.dataItem;
    };
    AgendaListComponent.prototype.formatTime = function (dataItem) {
        if (dataItem.isAllDay) {
            return this.localization.get('allDay');
        }
        var format = "{0:t}-{1:t}";
        if (dataItem.head) {
            format = "{0:t}";
        }
        else if (dataItem.tail) {
            format = "{1:t}";
        }
        return this.intlService.format(format, kendo_date_math_1.toLocalDate(dataItem.start), kendo_date_math_1.toLocalDate(dataItem.end));
    };
    AgendaListComponent.prototype.trackByFn = function (_, item) {
        return item.dataItem;
    };
    AgendaListComponent.decorators = [
        { type: core_1.Component, args: [{
                    // tslint:disable-next-line:component-selector
                    selector: '[kendoSchedulerAgendaList]',
                    template: "\n        <table class=\"k-scheduler-table\">\n            <tbody>\n                <ng-container *ngFor=\"let group of tasks; let groupIndex = index;\">\n                    <tr *ngFor=\"let item of group.tasks;let index = index; trackBy: trackByFn\">\n                        <ng-container *ngFor=\"let resource of group.resources;let resourceIndex = index\">\n                            <td *ngIf=\"group.spans[resourceIndex] && index === 0\" class=\"k-scheduler-groupcolumn\" [attr.rowspan]=\"group.spans[resourceIndex]\">\n                                {{ resource }}\n                            </td>\n                        </ng-container>\n                        <td *ngIf=\"item.type === 'group'\"\n                            [kendoSchedulerAgendaHeaderItem]=\"item\"\n                            [agendaDateTemplate]=\"agendaDateTemplate\">\n                        </td>\n                        <td class=\"k-scheduler-timecolumn\">\n                            <div *ngIf=\"!agendaTimeTemplate\">\n                                <span class=\"k-icon k-i-arrow-60-left\" *ngIf=\"extractDataItem(item).tail\"></span>\n                                {{formatTime(extractDataItem(item)) }}\n                                <span class=\"k-icon k-i-arrow-60-right\" *ngIf=\"extractDataItem(item).head\"></span>\n                            </div>\n                            <ng-container *ngIf=\"agendaTimeTemplate\" [ngTemplateOutlet]=\"agendaTimeTemplate\"\n                                [ngTemplateOutletContext]=\"extractDataItem(item)\">\n                            </ng-container>\n                        </td>\n                        <td [attr.data-group-index]=\"groupIndex\" [attr.data-task-index]=\"index\" [editable]=\"editable\"\n                            [kendoSchedulerAgendaTaskItem]=\"extractDataItem(item)\" [eventTemplate]=\"eventTemplate\"></td>\n                    </tr>\n                </ng-container>\n            </tbody>\n        </table>\n    "
                },] },
    ];
    /** @nocollapse */
    AgendaListComponent.ctorParameters = function () { return [
        { type: kendo_angular_intl_1.IntlService, },
        { type: kendo_angular_l10n_1.LocalizationService, },
    ]; };
    AgendaListComponent.propDecorators = {
        'classes': [{ type: core_1.HostBinding, args: ["class.k-scheduler-content",] },],
        'tasks': [{ type: core_1.Input },],
        'eventTemplate': [{ type: core_1.Input },],
        'agendaTimeTemplate': [{ type: core_1.Input },],
        'agendaDateTemplate': [{ type: core_1.Input },],
        'editable': [{ type: core_1.Input },],
    };
    return AgendaListComponent;
}());
exports.AgendaListComponent = AgendaListComponent;
