// @progress/kendo-angular-toolbar/dist/npm/navigation/tool-navigation.service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
/**
 * @hidden
 */
var ToolNavigationService = /** @class */ (function () {
    function ToolNavigationService() {
    }
    ToolNavigationService.prototype.register = function (rendererService, location) {
        if (location === 'toolbar') {
            this.toolbarRenderer = rendererService;
        }
        else {
            this.overflowRenderer = rendererService;
        }
    };
    ToolNavigationService.prototype.canFocus = function () {
        return false;
    };
    ToolNavigationService.prototype.focus = function () { };
    ToolNavigationService.prototype.defocus = function () { };
    ToolNavigationService.prototype.hasFocus = function () {
        return false;
    };
    ToolNavigationService.decorators = [
        { type: core_1.Injectable },
    ];
    /** @nocollapse */
    ToolNavigationService.ctorParameters = function () { return []; };
    return ToolNavigationService;
}());
exports.ToolNavigationService = ToolNavigationService;
