// @progress/kendo-angular-sortable/dist/es/sortable.service.js
import { Injectable, NgZone } from '@angular/core';
import Draggable from '@telerik/kendo-draggable';
import { draggableFromEvent, isDocumentAvailable } from './util';
import { Subject } from 'rxjs/Subject';
import { switchMap } from 'rxjs/operators/switchMap';
import { filter } from 'rxjs/operators/filter';
import { take } from 'rxjs/operators/take';
import { tap } from 'rxjs/operators/tap';
/**
 * The service that provides the drag-and-drop functionality for
 * transferring items between Sortable components within the same page.
 *
 */
var SortableService = /** @class */ (function () {
    function SortableService(ngZone) {
        var _this = this;
        this.ngZone = ngZone;
        /**
         * Specifies the Draggable item that is currently being moved.
         * @hidden
         */
        this.activeDraggable = null;
        /**
         * Specifies the Draggable item from which the dragging started.
         * @hidden
         */
        this.originDraggable = null;
        /**
         * @hidden
         */
        this.targetSortable = null;
        /**
         * Specifies the Draggable item that last emitted an event.
         * @hidden
         */
        this.lastDraggable = null;
        /**
         * @hidden
         */
        this.onPressSubject = new Subject();
        /**
         * @hidden
         */
        this.onDragSubject = new Subject();
        /**
         * @hidden
         */
        this.onReleaseSubject = new Subject();
        this.source = null;
        this._target = null;
        this.sortableCounter = 0;
        this.sortableRegister = {};
        if (!isDocumentAvailable()) {
            return;
        }
        this.draggable = new Draggable({
            drag: function (e) { return _this.onDragSubject.next(e); },
            press: function (e) { return _this.onPressSubject.next(e); },
            release: function (e) { return _this.onReleaseSubject.next(e); }
        });
        this.ngZone.runOutsideAngular(function () {
            _this.draggable.bindTo(document);
        });
        this.subscriptions = this.onPressSubject.pipe(tap(function (press) {
            _this.ngZone.run(function () {
                _this.targetSortable = _this.getSortableComponentFromTouch(press);
            });
        }), filter(function (_) { return !!_this.targetSortable; }), tap(function (press) {
            _this.onReleaseSubject.pipe(take(1)).subscribe(function (event) { return _this.release(event); });
            _this.start(press);
        }), switchMap(function (_drag) {
            return _this.onDragSubject.pipe(filter(function (_) { return !!_this.targetSortable; }), //stop further events if dragStart is prevented
            filter(function (_) { return _this.source !== null; }), tap(function (e) { return _this.drag(e); }));
        })).subscribe();
    }
    Object.defineProperty(SortableService.prototype, "target", {
        get: function () {
            return this._target;
        },
        /**
         * Specifies the `SortableComponent` instance under the dragged item.
         * @hidden
         */
        set: function (target) {
            this._target = target;
        },
        enumerable: true,
        configurable: true
    });
    SortableService.prototype.ngOnDestroy = function () {
        if (this.subscriptions) {
            this.subscriptions.unsubscribe();
        }
        if (this.draggable) {
            this.draggable.destroy();
        }
    };
    /**
     * Registers a `SortableComponent` with which the service operates.
     *
     * @param sortableComponent - The `SortableComponent`.
     * @return - The unique key that the current `SortableComponent` gets when registered.
     */
    SortableService.prototype.registerComponent = function (sortableComponent) {
        var id = this.sortableCounter.toString();
        this.sortableRegister[id] = sortableComponent;
        this.sortableCounter++;
        return id;
    };
    /**
     * Removes a `SortableComponent` from the registered `SortableComponents` with which the service operates.
     *
     * @param key - The key of the `SortableComponent` which will be removed from the register.
     * Obtained when `registerComponent` is called.
     */
    SortableService.prototype.unregisterComponent = function (key) {
        this.sortableRegister[key] = null;
    };
    /**
     * Sets the `SortableComponent` as a source component. When dragging an item from one Sortable to another,
     * the source component is the one from which the item originates.
     *
     * @param sortable - The `SortableComponent`.
     */
    SortableService.prototype.setSource = function (sortable) {
        this.source = sortable;
    };
    /**
     * Returns the source `SortableComponent` from which
     * an item is dragged to other Sortable components.
     *
     * @return - The `SourceComponent`.
     */
    SortableService.prototype.getSource = function () {
        return this.source;
    };
    /**
     * The method that finds the `SortableComponent` which is registered to
     * the `SortableService` by using the arguments of the `touch` event.
     *
     * @param touch - A Touch-Object of the `Touch` type interface.
     * Represents a single contact point (finger or stylus)
     * on a touch-sensitive device (touchscreen or trackpad).
     *
     * @return { component: SortableComponent, index: number } - An object
     * where the component is the `SortableComponent` that owns the item
     * and the index is the index of the touched item.
     */
    SortableService.prototype.getSortableComponentFromTouch = function (touch) {
        if (!isDocumentAvailable()) {
            return { component: undefined, index: undefined };
        }
        var realTarget = document.elementFromPoint(touch.clientX, touch.clientY);
        while (realTarget) {
            var id = realTarget.getAttribute('data-sortable-id');
            var index = realTarget.getAttribute('data-sortable-index');
            if (id) {
                var targetSortable = this.sortableRegister[id];
                if (targetSortable) {
                    return { component: targetSortable, index: parseInt(index, 10) };
                }
            }
            realTarget = realTarget.parentElement;
        }
    };
    SortableService.prototype.start = function (event) {
        var _this = this;
        this.ngZone.run(function () {
            var startTarget = draggableFromEvent(event, _this.targetSortable.component);
            if (_this.targetSortable.component.startDrag({ target: startTarget, originalEvent: event })) {
                _this.targetSortable = null;
            }
        });
    };
    SortableService.prototype.release = function (event) {
        var _this = this;
        this.ngZone.run(function () {
            if (!!_this.targetSortable) {
                var dropTarget = draggableFromEvent(event, _this.targetSortable.component);
                _this.source.endDrag({ target: dropTarget, originalEvent: event });
            }
            _this.source.positionHintFromEvent(null);
            _this.targetSortable = null;
        });
    };
    SortableService.prototype.drag = function (event) {
        var _this = this;
        this.ngZone.run(function () {
            _this.source.positionHintFromEvent(event);
            var sortable = _this.getSortableComponentFromTouch(event);
            if (!sortable || sortable && sortable.component !== _this.target) {
                if (_this.target) {
                    _this.target.leave({ target: undefined, originalEvent: event });
                }
                else if (_this.source !== _this.target) {
                    _this.source.leave({ target: undefined, originalEvent: event });
                }
            }
            if (sortable && sortable.component) {
                var draggable = draggableFromEvent(event, sortable.component);
                sortable.component.drag({ target: draggable, originalEvent: event });
            }
        });
    };
    SortableService.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    SortableService.ctorParameters = function () { return [
        { type: NgZone, },
    ]; };
    return SortableService;
}());
export { SortableService };
