// @progress/kendo-angular-sortable/dist/es/util.js
var NODE_NAME_PREDICATES = {};
var NODE_ATTR_PREDICATES = {};
/**
 * @hidden
 */
export var isDocumentAvailable = function () { return typeof document !== 'undefined'; };
/**
 * @hidden
 */
export var matchesNodeName = function (nodeName) {
    if (!NODE_NAME_PREDICATES[nodeName]) {
        NODE_NAME_PREDICATES[nodeName] = function (element) {
            return String(element.nodeName).toLowerCase() === nodeName.toLowerCase();
        };
    }
    return NODE_NAME_PREDICATES[nodeName];
};
/**
 * @hidden
 */
export var matchesNodeAttr = function (nodeAttr) {
    if (!NODE_ATTR_PREDICATES[nodeAttr]) {
        NODE_ATTR_PREDICATES[nodeAttr] = function (element) {
            return element.hasAttribute ? element.hasAttribute(nodeAttr) : false;
        };
    }
    return NODE_ATTR_PREDICATES[nodeAttr];
};
/**
 * @hidden
 */
export var closest = function (node, predicate) {
    while (node && !predicate(node)) {
        node = node.parentNode;
    }
    return node;
};
/**
 * Returns an object specifiying whether there is a DraggableDirective under the cursor.
 * @hidden
 */
export var draggableFromPoint = function (x, y) {
    if (!isDocumentAvailable()) {
        return;
    }
    var el = document.elementFromPoint(x, y);
    if (!el) {
        return;
    }
    var isDraggable = el.hasAttribute("kendoDraggable");
    var isChild = closest(el, matchesNodeAttr("kendoDraggable")) !== null;
    var parentDraggable = closest(el, matchesNodeAttr("kendoDraggable"));
    var index = parentDraggable ? parseInt(parentDraggable.getAttribute("data-sortable-index"), 10) : -1;
    return {
        element: el,
        index: index,
        isDraggable: isDraggable,
        isDraggableChild: isChild,
        parentDraggable: parentDraggable,
        rect: el.getBoundingClientRect()
    };
};
/**
 * Returns the DraggableDirective under the cursor.
 * @hidden
 */
export var draggableFromEvent = function (event, sortable) {
    var target;
    if (event.changedTouches) {
        var touch = event.changedTouches[0];
        target = draggableFromPoint(touch.clientX, touch.clientY);
    }
    else {
        target = draggableFromPoint(event.clientX, event.clientY);
    }
    // TODO: refactor sortable. Add draggable getter
    return sortable.draggables.toArray()[target ? target.index : -1];
};
