// @progress/kendo-angular-editor/dist/es/tools/editor-button.directive.js
import { Directive, Input, Host } from '@angular/core';
import { ToolBarButtonComponent } from '@progress/kendo-angular-toolbar';
import { commandIcons } from '../config/command-icons';
import { EditorComponent } from '../editor.component';
/**
 * A directive which configures an existing `ToolBarButtonComponent` as an Editor tool
 * ([see example]({% slug toolbartools_editor %}#toc-built-in-tools)). According to the command that
 * is passed to it, `kendoEditorButton` predefines the `icon` and `click` event handlers of the button.
 * In addition, the directive updates the `selected` or `disabled` state of the button while the user is editing.
 *
 * @example
 * ```ts-no-run
 * <kendo-toolbar-button kendoEditorButton="bold"></kendo-toolbar-button>
 * ```
 *
 * The following example demonstrates how to change the default icon applied by the directive.
 *
 * @example
 * ```ts-no-run
 * <kendo-toolbar-button kendoEditorButton="bold" icon="blogger"></kendo-toolbar-button>
 * ```
 */
var EditorButtonDirective = /** @class */ (function () {
    function EditorButtonDirective(button, editor) {
        this.button = button;
        this.editor = editor;
    }
    EditorButtonDirective.prototype.ngOnInit = function () {
        var _this = this;
        if (!this.button.icon) {
            this.button.icon = commandIcons[this.command];
        }
        this.subs = this.button.click.subscribe(function () {
            _this.editor.exec(_this.command);
            _this.editor.focus();
        });
        this.subs.add(this.editor.stateChange.subscribe(function (state) {
            if (_this.command === 'indent' || _this.command === 'outdent') {
                _this.button.disabled = !state[_this.command];
            }
            else if (_this.command === 'unlink') {
                _this.button.disabled = !state.link;
            }
            else if (_this.command === 'undo' || _this.command === 'redo') {
                _this.button.disabled = state[_this.command];
            }
            else {
                _this.button.selected = state[_this.command];
            }
        }));
    };
    EditorButtonDirective.prototype.ngOnDestroy = function () {
        this.subs.unsubscribe();
    };
    EditorButtonDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'kendo-toolbar-button[kendoEditorButton]'
                },] },
    ];
    /** @nocollapse */
    EditorButtonDirective.ctorParameters = function () { return [
        { type: ToolBarButtonComponent, },
        { type: EditorComponent, decorators: [{ type: Host },] },
    ]; };
    EditorButtonDirective.propDecorators = {
        'command': [{ type: Input, args: ['kendoEditorButton',] },],
    };
    return EditorButtonDirective;
}());
export { EditorButtonDirective };
