// @progress/kendo-angular-editor/dist/es/dialogs/image-dialog.component.js
import * as tslib_1 from "tslib";
import { Component, Input } from '@angular/core';
import { FormControl, FormGroup, Validators } from '@angular/forms';
import { DialogRef, DialogContentBase } from '@progress/kendo-angular-dialog';
import { getNodeFromSelection } from '../prosemirror-utils';
import { safeString } from '../util';
/**
 * @hidden
 */
var ImageDialogComponent = /** @class */ (function (_super) {
    tslib_1.__extends(ImageDialogComponent, _super);
    function ImageDialogComponent(dialog) {
        var _this = _super.call(this, dialog) || this;
        _this.dialog = dialog;
        _this.data = {
            src: '',
            alt: '',
            width: '',
            height: ''
        };
        _this.src = new FormControl('', Validators.required);
        _this.alt = new FormControl('');
        _this.width = new FormControl('', Validators.min(1));
        _this.height = new FormControl('', Validators.min(1));
        _this.imageData = new FormGroup({
            src: _this.src,
            alt: _this.alt,
            width: _this.width,
            height: _this.height
        });
        return _this;
    }
    ImageDialogComponent.prototype.onCancelAction = function () {
        this.dialog.close();
    };
    ImageDialogComponent.prototype.onConfirmAction = function () {
        if (this.src.value) {
            this.editor.exec('insertImage', this.getData());
            this.dialog.close();
            this.editor.focus();
        }
    };
    ImageDialogComponent.prototype.setData = function (state) {
        var node = getNodeFromSelection(state);
        if (node) {
            this.src.patchValue(node.attrs.src);
            this.alt.patchValue(node.attrs.alt);
            this.width.patchValue(node.attrs.width);
            this.height.patchValue(node.attrs.height);
        }
    };
    ImageDialogComponent.prototype.getData = function () {
        return {
            src: this.src.value,
            alt: this.alt.value,
            width: this.normalizeDimension(this.width.value),
            height: this.normalizeDimension(this.height.value)
        };
    };
    ImageDialogComponent.prototype.normalizeDimension = function (value) {
        return Number.isNaN(parseInt(value, 10)) || parseInt(value, 10) <= 0 ? '' : safeString(parseInt(value, 10));
    };
    ImageDialogComponent.decorators = [
        { type: Component, args: [{
                    template: "\n        <kendo-dialog-titlebar>\n            Insert Image\n        </kendo-dialog-titlebar>\n        <div class=\"k-editor-dialog\">\n            <div class=\"k-editor-dialog k-popup-edit-form k-window-content k-content\" >\n                <div class=\"k-edit-form-container k-window-content\">\n                    <div class=\"k-edit-label\">\n                        <label for=\"k-editor-image-url\">Web address</label>\n                    </div>\n                    <div class=\"k-edit-field\">\n                        <input [formControl]=\"src\" type=\"text\" class=\"k-input k-textbox\">\n                    </div>\n                    <div class=\"k-edit-label\">\n                        <label for=\"k-editor-image-text\">Alternate text</label>\n                    </div>\n                    <div class=\"k-edit-field\">\n                        <input [formControl]=\"alt\" type=\"text\" class=\"k-input k-textbox\">\n                    </div>\n                    <div class=\"k-edit-label\">\n                        <label for=\"k-editor-image-width\">Width (px)</label>\n                    </div>\n                    <div class=\"k-edit-field\">\n                        <input [formControl]=\"width\" type=\"text\" class=\"k-input k-textbox\">\n                    </div>\n                    <div class=\"k-edit-label\">\n                        <label for=\"k-editor-image-height\">Height (px)</label>\n                    </div>\n                    <div class=\"k-edit-field\">\n                        <input [formControl]=\"height\" type=\"text\" class=\"k-input k-textbox\">\n                    </div>\n                </div>\n            </div>\n        </div>\n        <kendo-dialog-actions>\n            <button kendoButton (click)=\"onCancelAction()\">Cancel</button>\n            <button kendoButton [disabled]=\"imageData.invalid\" (click)=\"onConfirmAction()\" [primary]=\"true\">Insert</button>\n        </kendo-dialog-actions>\n    "
                },] },
    ];
    /** @nocollapse */
    ImageDialogComponent.ctorParameters = function () { return [
        { type: DialogRef, },
    ]; };
    ImageDialogComponent.propDecorators = {
        'editor': [{ type: Input },],
    };
    return ImageDialogComponent;
}(DialogContentBase));
export { ImageDialogComponent };
