// @progress/kendo-angular-editor/dist/es/dialogs/source-dialog.component.js
import * as tslib_1 from "tslib";
import { Component, ViewChild, Input } from '@angular/core';
// import { EditorService } from '../editor.service';
import { DialogRef, DialogContentBase } from '@progress/kendo-angular-dialog';
/**
 * @hidden
 */
var SourceDialogComponent = /** @class */ (function (_super) {
    tslib_1.__extends(SourceDialogComponent, _super);
    function SourceDialogComponent(dialog) {
        var _this = _super.call(this, dialog) || this;
        _this.dialog = dialog;
        _this.data = '';
        return _this;
    }
    SourceDialogComponent.prototype.onCancelAction = function () {
        this.dialog.close();
    };
    SourceDialogComponent.prototype.onConfirmAction = function () {
        this.editor.exec('setHTML', this.getData());
        this.dialog.close();
        this.editor.focus();
    };
    SourceDialogComponent.prototype.getData = function () {
        return this.textarea.nativeElement.value;
    };
    SourceDialogComponent.prototype.setData = function () {
        this.data = this.indent(this.editor.getSource());
    };
    SourceDialogComponent.prototype.indent = function (content) {
        return content.replace(/<\/(p|li|ul|ol|h[1-6]|table|tr|td|th)>/ig, "</$1>\n")
            .replace(/<(ul|ol)([^>]*)><li/ig, "<$1$2>\n<li")
            .replace(/<br \/>/ig, "<br />\n")
            .replace(/\n$/, "");
    };
    SourceDialogComponent.decorators = [
        { type: Component, args: [{
                    template: "\n        <kendo-dialog-titlebar>\n            View Source\n        </kendo-dialog-titlebar>\n        <textarea [value]=\"data\" #textarea class=\"k-editor-textarea k-input\"></textarea>\n        <kendo-dialog-actions>\n            <button kendoButton (click)=\"onCancelAction()\">Cancel</button>\n            <button kendoButton (click)=\"onConfirmAction()\" [primary]=\"true\">Update</button>\n        </kendo-dialog-actions>\n    ",
                    styles: ["\n        >>> .k-editor-textarea {\n            height: 100%;\n        }\n    "]
                },] },
    ];
    /** @nocollapse */
    SourceDialogComponent.ctorParameters = function () { return [
        { type: DialogRef, },
    ]; };
    SourceDialogComponent.propDecorators = {
        'editor': [{ type: Input },],
        'textarea': [{ type: ViewChild, args: ['textarea',] },],
    };
    return SourceDialogComponent;
}(DialogContentBase));
export { SourceDialogComponent };
