// @progress/kendo-angular-scheduler/dist/es2015/views/month/month-view.component.js
import { Component, ContentChild, forwardRef, Input, ChangeDetectionStrategy, ChangeDetectorRef } from '@angular/core';
import { LocalizationService } from '@progress/kendo-angular-l10n';
import { SchedulerView } from '../../types';
import { ConfigurationViewBase } from '../common/configuration-view-base';
import { ViewContextService } from '../view-context.service';
import { MonthDaySlotTemplateDirective } from '../templates';
import { DEFAULT_EVENT_HEIGHT } from '../constants';
import { isPresent } from '../../common/util';
/**
 * The component for rendering the **Month** view.
 */
export class MonthViewComponent extends ConfigurationViewBase {
    constructor(localization, viewContext, changeDetector) {
        super(localization, changeDetector);
        this.viewContext = viewContext;
        /**
         * The long-date format for displaying the
         * selected date in the Scheduler toolbar.
         * Defaults to `{0:Y}`
         * ([more information]({% slug parsingandformatting_intl %}#toc-date-formatting)).
         */
        this.selectedDateFormat = '{0:Y}';
        /**
         * The short-date format for displaying the
         * selected date in the Scheduler toolbar.
         * Defaults to `{0:y}`
         * ([more information]({% slug parsingandformatting_intl %}#toc-date-formatting.
         */
        this.selectedShortDateFormat = '{0:y}';
        /**
         * The invariant name for this view (`month`).
         */
        this.name = 'month';
        this.optionsChangeSubscription = this.viewContext.optionsChange.subscribe(this.optionsChange.bind(this));
    }
    /**
     * @hidden
     */
    get title() {
        return this.localization.get('monthViewTitle');
    }
    get viewEventHeight() {
        return isPresent(this.eventHeight) ? this.eventHeight : this.schedulerEventHeight;
    }
    ngOnDestroy() {
        this.optionsChangeSubscription.unsubscribe();
    }
    optionsChange(options) {
        this.schedulerEventHeight = options.eventHeight || DEFAULT_EVENT_HEIGHT;
    }
}
MonthViewComponent.decorators = [
    { type: Component, args: [{
                changeDetection: ChangeDetectionStrategy.OnPush,
                selector: 'kendo-scheduler-month-view',
                providers: [{
                        provide: SchedulerView,
                        // tslint:disable-next-line:no-forward-ref
                        useExisting: forwardRef(() => MonthViewComponent)
                    }],
                template: `
        <ng-template #content>
            <month-view
                [eventHeight]="viewEventHeight"
                [еventTemplate]="eventTemplate?.templateRef"
                [groupHeaderTemplate]="groupHeaderTemplate?.templateRef"
                [monthDaySlotTemplate]="monthDaySlotTemplate?.templateRef"
                [selectedDateFormat]="selectedDateFormat"
                [selectedShortDateFormat]="selectedShortDateFormat">
            </month-view>
        </ng-template>
    `
            },] },
];
/** @nocollapse */
MonthViewComponent.ctorParameters = () => [
    { type: LocalizationService, },
    { type: ViewContextService, },
    { type: ChangeDetectorRef, },
];
MonthViewComponent.propDecorators = {
    'eventHeight': [{ type: Input },],
    'selectedDateFormat': [{ type: Input },],
    'selectedShortDateFormat': [{ type: Input },],
    'monthDaySlotTemplate': [{ type: ContentChild, args: [MonthDaySlotTemplateDirective,] },],
};
